Remove tf.contrib dependency

adrianc-a opened the pull request:
Removing dependency on `tf.contrib` as part of a larger Google Summer of Code project to make tf-slim compatible with tensorflow 2.0.

### Changes
- `tf.contrib.layers`, `tf.contrib.framework.arg_scope`, `tf.contrib.framework.variable` `tf.contrib.training.training`, `tf.contrib.evaluation` moved into this repo. 
- `tf.contrib.losses` removed, used `tf.losses` -- might want to change this since `__init__.py` exports `tf.contrib.losses`
- `tf.contrib.util.deprecated` removed -- TODO replace with `tensorflow.python.util.deprecation`
- `tf.contrib.util.loader` replaced with `tf.load_op_library` -- TODO remove custom ops as per recommendation by Tomer. 

### Remaining work
- [x] Unit testing
- [x] Pip install 
- [x] Custom op removal -- Removing `sparse_feature_cross` causes many issues in `embedding_ops.py` and `layers/feature_column.py`
- [x] Re-deprecate deprecated functions in `layers/target_column.py`

Intended work to be completed after GSoC (final commit 31319935bc2c67688f52761b34251e9f66e2a3fd)

adrianc-a commented:
I've signed it.

sguada commented on tests_release.sh:
```diff
@@ -37,18 +34,18 @@ run_tests() {
   # TensorFlow isn't a regular dependency because there are many different pip
   # packages a user might have installed.
-  pip install tensorflow
+  pip install tensorflow==2.0.0-beta1
```
Not sure if we should pin to an specific beta version.

adrianc-a replied:
I agree. How should I proceed? 

Not specifying a version at all leads to installing r1.13/r1.14. There are different tests which fail when using one of these two versions of tensorflow (compare with r2.0). Should I have some way of version checking in these tests that differ?

Not specifying a beta version seems incompatible with pip, which wants either b1 or b0 for r2.0.

adrianc-a replied:
Changed in 90a8f81 to just install the latest version of tf.

sguada commented on tf_slim/training/__init__.py:
```diff
@@ -0,0 +1,17 @@
+# limitations under the License.
+# ==============================================================================
+from tf_slim.training import evaluation
+from tf_slim.training import training
```
Add extra empty line.

adrianc-a replied:
fixed

sguada commented on tf_slim/__init__.py:
```diff
@@ -48,4 +48,8 @@
 
+
+from tensorflow.compat.v1 import disable_eager_execution
+disable_eager_execution()
```
Do we need to disable_eager_execution in the init? or is enough in the tests?

adrianc-a replied:
It's enough in some (though nearly all) tests. [Here is the change](https://github.com/adrianc-a/tf-slim/tree/no_eager_in_tests). Let me know if I should merge it into this branch.

tomerk replied:
We want to avoid calling disable_eager_execution within libraries, because it can silently leak and cause issues. (Plus it's weird to have an import change your TensorFlow runtime settings.

My recommendation would be to create a utility method that looks something like:
```
def error_if_eager_is_enabled():
  with tf.init_scope():
    if tf.executing_eagerly():
      raise RuntimeError('TF-slim layers cannot be used when eager execution is enabled. Make sure to disable eager execution by calling `tf.compat.v1.disable_eager_execution()` or by using a custom estimator model_fn.')
```

And then run it at the start of every layer.
Then, all the tests should make sure to disable eager execution (but there could also be one test that makes sure the error gets raised if eager hasn't been disabled).

dubesar commented:
@sguada  why is this pr still not merged as all the changes look fine?

adrianc-a commented:
In its current state, this PR is not ready to be merged. 

Most importantly some tests still fail which I think requires some digging. @sguada mentioned he would comment to assist on this at some point. I haven't had time to work on this recently, but there is some low-hanging fruit w.r.t. to Sergio's earlier comments I can maybe fix soon since tf-2.0 has an official release now.

dubesar commented:
@adrianc-a if possible can I help to fix the parts remaining?

tomerk commented:
Hi @dubesar, if you're interested in helping @adrianc-a dig into the failing tests that would be much appreciated.

I should be able to help review the PR once the tests are working (or possibly to provide some help in figuring out the failing test cases).

pichuan commented:
Hi @tomerk , @adrianc-a ,
I made some suggestions in this pull request:
https://github.com/adrianc-a/tf-slim/pull/1

Here is how I confirmed that the tests passed:

```
gcloud  compute instances create “pichuan-test" --image-family "ubuntu-1604-lts"  --image-project "ubuntu-os-cloud"  --machine-type "n1-standard-8"    --zone "us-west1-b"  
gcloud compute ssh     pichuan-test

# On the GCE instance:
git clone https://github.com/pichuan/tf-slim.git
cd tf-slim
git checkout remove_contrib_tests_passed
sudo apt -y update
sudo apt-get -y install python-dev python-pip python-wheel
python -m pip install --user virtualenv
sudo add-apt-repository -y ppa:jonathonf/python-3.6
sudo apt -y update
sudo apt-get install -y python3.6
./tests_release.sh 2>&1 | tee /tmp/log
```

Let me know what you think. A few asserts changed there. I'm not exactly sure the new semantics make sense or not.

adrianc-a commented:
I've merged @pichuan 's changes into my fork so as to shift the discussion here (Thanks again for your PR). To me they look fine, though maybe @tomerk should also take a look. Specifically, `optimizers_test#testGradientNoise` and `testGradientNoiseWithClipping` as I was unsure what to do with them before. 

I confirm that they do fix all the failing tests. 

There are at least two other issues brought up by @sguada that should be solved/decided as well:

1. Whether to use `disable_eager_execution` in `__init__.py` or to only use it in the tests when required (in many). [The change is here](https://github.com/adrianc-a/tf-slim/tree/no_eager_in_tests) and I can merge into this PR should someone like. 

2. `tf.contrib.losses` I removed and replaced all uses with `tf.losses`. However, it's probably more reasonable to move `tf.contrib.losses` into this repo since slim exported these losses.

If there are any other issues that I've forgotten please let me know.

tomerk commented on tf_slim/layers/optimizers_test.py:
```diff
@@ -0,0 +1,540 @@
+        optimizers_lib.optimize_loss(
+            loss, global_step, learning_rate=-0.1, optimizer="SGD")
+
+  def testGradientNoise(self):
```
This and the below test seem like they're going to be very flaky and not super reliable.

If you have the time I'd recommend replacing it w/ a test that runs multiple times & checks the mean & std dev:

Reset variables to a fixed constant value after each run, & run it multiple times then test the mean & std dev of the results to make sure if it matches what you expect.

This isn't super high priority though. Since this test is already not great, I'd be okay w/ just deleting it in this PR and coming back to it later if you have the time. (I think it's better to remove this test than leave it in, because it's not a great signal and is likely to break)

pichuan commented:
@adrianc-a let me know if you're going to work on the suggestions that @tomerk had above, or if you prefer me to take some of them.

adrianc-a commented:
@pichuan I can work on 3 and 1 if you want to help out on 2 that would be great. Thanks for helping out.

If you are on some deadline then let me know and you can take over more work since I have other things to attend to at the moment.

pichuan commented:
@adrianc-a I'll add 2 to my TODO list. This is not blocking me because right now I can just pip install a branch that I made. It would still be nice for this to be merged though.
I also have a bunch of other stuff on my TODO list. If I start looking into 2, I'll add another comment here. (Otherwise you can assume that I have not started)
Thanks!

pichuan commented:
Following up on my last comment, I started looking into @tomerk 's suggestion on adding tf.contrib.losses. What I'm doing now is trying to directly copy code for tf.contrib.losses into tf_slim, so that tf_slim can continue to use tf_slim.losses as an alias.
Right now, I have a bunch of test errors that I will try to address.

If this is not what's expected for me to do, let me know soon. 
When/if I'm able to make tests pass again, I'll send another PR for @adrianc-a 's branch. I have no idea how long it'll take me. But I have started the work.

adrianc-a commented:
@sguada @pichuan was this merge an error/mistake? I unfortunately have been busy and haven't done the tasks I mentioned above and I don't see @pichuan 's changes either. As such, isn't this pr still unfinished?

adrianc-a commented:
You can disregard my deleted comment you may have been notified by. 

Thanks @sguada and @tfboyd for the commits/merge.

However, I pulled from master and cannot run the tests due to some import issues. Specifically with `tf_sim.ops` and more specifically with `tf_slim.ops.arg_scope`. Not sure if it's something on my side but I thought to bring it up. 

Running from this branch on r2.0.0 is fine but running on r2.1.0 results in a test failure (`tf_slim.layers.layers_test.LegacyFullyConnectedTest`) as well.
